src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Services/InMemoryCacheServiceTests.cs

```diff
@@ -0,0 +1,441 @@
+                    59, 
+                    30)
+                .ToUniversalTime();
+
+        // Set the expected cache duration to be 45 seconds as per the requested duration.
+        // This should be truncated as it carries 15 seconds over into the next half hour.
+        var expectedCacheExpiry = now.AddSeconds(45 - 15);
+
+        await SetItemAndAssertExpiryTime(now, cacheConfiguration, expectedCacheExpiry);
+    }
+        
+    [Fact]
+    public async Task SetItem_ExpiryTimeNotTruncated_NoExpirySchedule()
+    {
+        // The requested ExpirySchedule is hourly, meaning that cached items will have their 
+        // expiry times truncated if the requested cache duration would carry over into a new
+        // hour.
+        var cacheConfiguration = new InMemoryCacheConfiguration(
+            ExpirySchedule.None, 
+            6000);
+
+        var now = 
+            new DateTime(
+                    2022, 
+                    07, 
+                    18, 
+                    0, 
+                    29, 
+                    30)
+                .ToUniversalTime();
+
+        // The requested cache duration is really long, but because there's no regular cache clearing
+        // schedule, the requested cache duration will be honoured.
+        var expectedCacheExpiry = now.AddSeconds(6000);
+
+        await SetItemAndAssertExpiryTime(now, cacheConfiguration, expectedCacheExpiry);
+    }
+
+    private async Task SetItemAndAssertExpiryTime(
+        DateTime now,
+        InMemoryCacheConfiguration cacheConfiguration,
+        DateTime expectedCacheExpiry)
+    {
+        var cacheKey = new SampleCacheKey("Key");
+        const string valueToCache = "test item";
+
+        var memoryCache = new Mock<IMemoryCache>(Strict);
+        var cacheEntry = new Mock<ICacheEntry>(Strict);
+
+        // Options that we expect to see set.
+        cacheEntry
+            .SetupSet(s => s.Size = $"\"{valueToCache}\"".Length);
+
+        cacheEntry
+            .SetupSet(s => s.AbsoluteExpiration = new DateTimeOffset(expectedCacheExpiry));
+
+        // Default unset options.
+        cacheEntry.SetupSet(s => s.AbsoluteExpirationRelativeToNow = null);
+        cacheEntry.SetupSet(s => s.SlidingExpiration = null);
+        cacheEntry.SetupSet(s => s.Priority = CacheItemPriority.Normal);
+        cacheEntry.SetupSet(s => s.Value = valueToCache);
+        cacheEntry.Setup(s => s.Dispose());
+            
+        memoryCache
+            .Setup(mock => mock.CreateEntry(cacheKey))
+            .Returns(() => cacheEntry.Object);
+
+        var service = SetupService(memoryCache.Object);
+        await service.SetItem(cacheKey, valueToCache, cacheConfiguration, now);
+            
+        VerifyAllMocks(memoryCache, cacheEntry);
+    }
+
+    [Fact]
+    public async Task GetItem()
+    {
+        object entity = new SampleClass();
+
+        var cacheKey = new SampleCacheKey("Key");
+
+        var memoryCache = new Mock<IMemoryCache>(Strict);
```

Comment:
I must admit, it hadn't occurred to me to do it differently before, and actually mocking the IMemoryCache code is a right pain because of all the Extension methods that we actually use are not mockable, so I was totally up for this suggestion!

Unfortunately, I've looked into it and I don't think I can verify what I want to verify without mocking the whole lot.  The problem stems from the fact that I want to verify the cache expiry options that are used against the real IMemoryCache instance, but this value isn't used as a parameter to any method on IMemoryCache itself, but is rather used *after* an `ICacheEntry` has been created using `cache.CreateEntry()`.

The problem stems from this extension code in `MemoryCacheExtensions.cs`:

```
        public static TItem Set<TItem>(this IMemoryCache cache, object key, TItem value, MemoryCacheEntryOptions options)
        {
            using ICacheEntry entry = cache.CreateEntry(key);
            if (options != null)
            {
                entry.SetOptions(options);
            }

            entry.Value = value;

            return value;
        }
```

Reply:
You can see here then that in order to see what (if anything) is done with the `options`, I need to somehow get a handle on the ICacheEntry returned from `cache.CreateEntry()` and then ensure the options are set on it properly.

If I could create a Spy that could capture the return value of a "real" `cache.CreateEntry()` call, this'd do, and I could just assert that the options are set on it correctly.  But Moq doesn't seem to support anything like this.  No Spies or ability to capture the return values from any methods where `CallBase()` is used.

So as it goes, I think I *have* to mock CreateEntry().

However, I think I can create a simple implementation of ICacheEntry rather than needing to mock that as well, which might then make the cache configuration expectation code a little nicer at least.

Reply:
But yeah, we can lose the Mocks for all the non `SetItem` tests, for sure!